File: Books/A_Comprehensive_Approach_to_ML_and_AI/python/rnn_lstm_advanced_tutorial.py
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from tensorflow.keras import layers, models, optimizers
from tensorflow.keras.callbacks import EarlyStopping

# 1. Load and Parse the Data
filename = 'data/jena_climate_2009_2016.csv'
df = pd.read_csv(filename)
# Assume the first column is date and the third column is temperature.
df['Date'] = pd.to_datetime(df.iloc[:,0])
temp = df.iloc[:,2].values  # temperature in °C
print(f"Loaded {len(temp)} samples from {df['Date'].iloc[0]} to {df['Date'].iloc[-1]}")

# 2. Plot Temperature Time-Series
plt.figure(figsize=(10,4))
plt.plot(df['Date'], temp)
plt.xlabel("Time")
plt.ylabel("Temperature (°C)")
plt.title("Temperature Time-Series")
plt.show()

# 3. Prepare Data: windowing function.
def create_dataset(data, lookback, delay, step, start_index, end_index):
    num_samples = (end_index - start_index - lookback - delay) + 1
    X = np.zeros((num_samples, lookback // step))
    y = np.zeros((num_samples,))
    for i in range(num_samples):
        indices = np.arange(start_index+i, start_index+i+lookback, step)
        X[i] = data[indices]
        y[i] = data[start_index+i+lookback+delay-1]
    return X, y

lookback = 720   # past 720 timesteps
delay = 144      # predict 144 timesteps ahead
step = 6         # sample every 6 timesteps

num_data = len(temp)
train_end = 200000
val_end = 300000
if num_data < train_end + lookback + delay:
    raise ValueError("Not enough data for these parameters.")

X_train, y_train = create_dataset(temp, lookback, delay, step, 0, train_end)
X_val, y_val = create_dataset(temp, lookback, delay, step, train_end, val_end)
X_test, y_test = create_dataset(temp, lookback, delay, step, val_end, num_data)

print("Training samples:", X_train.shape[0])
print("Validation samples:", X_val.shape[0])
print("Testing samples:", X_test.shape[0])

# 5. Normalize the Data using training statistics.
train_mean = X_train.mean()
train_std = X_train.std()
X_train = (X_train - train_mean) / train_std
X_val = (X_val - train_mean) / train_std
X_test = (X_test - train_mean) / train_std

y_train = (y_train - train_mean) / train_std
y_val = (y_val - train_mean) / train_std
y_test = (y_test - train_mean) / train_std

# 6. Baseline: predict last value.
y_pred_baseline = X_val[:,-1]
baseline_mae = np.mean(np.abs(y_pred_baseline - y_val))
print(f"Baseline MAE (normalized): {baseline_mae:.4f}")

# 7. Build models using Keras.
input_shape = (X_train.shape[1],)
def build_dense_model():
    model = models.Sequential([
        layers.Input(shape=input_shape),
        layers.Dense(32, activation='relu'),
        layers.Dense(1)
    ])
    model.compile(optimizer=optimizers.Adam(1e-3), loss='mse', metrics=['mae'])
    return model

def build_rnn_model():
    model = models.Sequential([
        layers.Input(shape=(X_train.shape[1], 1)),
        layers.SimpleRNN(32, return_sequences=False),
        layers.Dense(1)
    ])
    model.compile(optimizer=optimizers.Adam(1e-3), loss='mse', metrics=['mae'])
    return model

def build_lstm_model():
    model = models.Sequential([
        layers.Input(shape=(X_train.shape[1], 1)),
        layers.LSTM(32),
        layers.Dense(1)
    ])
    model.compile(optimizer=optimizers.Adam(1e-3), loss='mse', metrics=['mae'])
    return model

# Reshape data for RNN/LSTM (samples, timesteps, 1)
X_train_rnn = X_train[..., np.newaxis]
X_val_rnn = X_val[..., np.newaxis]
X_test_rnn = X_test[..., np.newaxis]

# Train Dense Model
dense_model = build_dense_model()
es = EarlyStopping(monitor='val_loss', patience=3)
history_dense = dense_model.fit(X_train, y_train, epochs=10, batch_size=32,
                                validation_data=(X_val, y_val), callbacks=[es], verbose=2)
y_pred_dense = dense_model.predict(X_val).flatten()
dense_mae = np.mean(np.abs(y_pred_dense - y_val))
print(f"Dense model MAE (normalized): {dense_mae:.4f}")

# Train RNN Model
rnn_model = build_rnn_model()
history_rnn = rnn_model.fit(X_train_rnn, y_train, epochs=10, batch_size=32,
                            validation_data=(X_val_rnn, y_val), callbacks=[es], verbose=2)
y_pred_rnn = rnn_model.predict(X_val_rnn).flatten()
rnn_mae = np.mean(np.abs(y_pred_rnn - y_val))
print(f"RNN model MAE (normalized): {rnn_mae:.4f}")

# Train LSTM Model
lstm_model = build_lstm_model()
history_lstm = lstm_model.fit(X_train_rnn, y_train, epochs=10, batch_size=32,
                              validation_data=(X_val_rnn, y_val), callbacks=[es], verbose=2)
y_pred_lstm = lstm_model.predict(X_val_rnn).flatten()
lstm_mae = np.mean(np.abs(y_pred_lstm - y_val))
print(f"LSTM model MAE (normalized): {lstm_mae:.4f}")

# 10. Plot predictions on validation set.
t = np.arange(len(y_val))
plt.figure(figsize=(10,6))
plt.plot(t, y_val, 'k', linewidth=1.5, label='True')
plt.plot(t, y_pred_baseline, 'b', label='Baseline')
plt.plot(t, y_pred_dense, 'r', label='Dense')
plt.plot(t, y_pred_rnn, 'g', label='RNN')
plt.plot(t, y_pred_lstm, 'm', label='LSTM')
plt.xlabel("Validation Sample Index")
plt.ylabel("Normalized Temperature")
plt.legend()
plt.title("Model Predictions on Validation Set")
plt.grid(True)
plt.show()

File: Books/A_Comprehensive_Approach_to_ML_and_AI/R/rnn_lstm_advanced_tutorial.R
# Advanced Climate Forecasting with Dense, RNN, and LSTM Models in R
#
# This tutorial uses the Jena climate dataset to forecast temperature.
# It performs the following steps:
#   1. Loads and parses the data.
#   2. Windows the data to create training, validation, and test sets.
#   3. Normalizes the data using training-set statistics.
#   4. Computes a baseline MAE.
#   5. Trains a densely connected model, an RNN model, and an LSTM model using keras.
#   6. Plots and compares predictions on the validation set.
#
# Note: This example uses the keras package. Install and configure keras beforehand.

library(keras)
library(tidyverse)

# 1. Load and Parse the Data
filename <- "data/jena_climate_2009_2016.csv"
if (!file.exists(filename)) stop("File not found. Please download jena_climate_2009_2016.csv.")
data <- read_csv(filename)
dt <- as.POSIXct(data[[1]], format="%Y-%m-%d %H:%M:%S")
temp <- data[[3]]
cat(sprintf("Loaded %d temperature samples from %s to %s\n", length(temp), as.character(dt[1]), as.character(dt[length(dt)])))

# 2. Prepare the Data (Windowing)
lookback <- 720  # number of past timesteps (~5 days if 10-min intervals)
delay <- 144     # predict 144 timesteps ahead (~1 day)
step <- 6        # sample every 6 timesteps (hourly)
start_index <- 1
end_index <- 200000  # adjust based on data length

create_dataset <- function(data, lookback, delay, step, start_index, end_index) {
  num_samples <- floor((end_index - start_index - lookback - delay) / 1) + 1
  num_timesteps <- floor(lookback / step)
  X <- array(0, dim = c(num_samples, num_timesteps, 1))
  y <- numeric(num_samples)
  for (i in 1:num_samples) {
    idx <- start_index + i - 1
    indices <- seq(idx, idx + lookback - 1, by = step)
    X[i,,1] <- data[indices]
    y[i] <- data[idx + lookback + delay - 1]
  }
  list(X = X, y = y)
}

dataset <- create_dataset(temp, lookback, delay, step, start_index, end_index)
X_train <- dataset$X
y_train <- dataset$y

# For validation and test, use later parts of the data.
val_dataset <- create_dataset(temp, lookback, delay, step, 200001, 300000)
X_val <- val_dataset$X
y_val <- val_dataset$y

# 3. Normalize the Data using training statistics
train_mean <- mean(X_train)
train_std <- sd(X_train)
X_train <- (X_train - train_mean) / train_std
X_val <- (X_val - train_mean) / train_std
y_train <- (y_train - train_mean) / train_std
y_val <- (y_val - train_mean) / train_std

# 4. Baseline: Predict last value in window as forecast.
y_pred_baseline <- X_val[, dim(X_val)[2], 1]
baseline_MAE <- mean(abs(y_pred_baseline - y_val))
cat(sprintf("Baseline MAE (normalized): %.4f\n", baseline_MAE))

# 5. Define and Train Models using Keras

# Dense Model
dense_model <- keras_model_sequential() %>%
  layer_flatten(input_shape = dim(X_train)[2:3]) %>%
  layer_dense(units = 32, activation = "relu") %>%
  layer_dense(units = 1)
dense_model %>% compile(optimizer = optimizer_adam(lr = 1e-3), loss = "mse")
history_dense <- dense_model %>% fit(X_train, y_train, epochs = 10, batch_size = 32,
                                      validation_data = list(X_val, y_val))
y_pred_dense <- dense_model %>% predict(X_val)

# RNN Model
rnn_model <- keras_model_sequential() %>%
  layer_simple_rnn(units = 32, input_shape = dim(X_train)[2:3]) %>%
  layer_dense(units = 1)
rnn_model %>% compile(optimizer = optimizer_adam(lr = 1e-3), loss = "mse")
history_rnn <- rnn_model %>% fit(X_train, y_train, epochs = 10, batch_size = 32,
                                 validation_data = list(X_val, y_val))
y_pred_rnn <- rnn_model %>% predict(X_val)

# LSTM Model
lstm_model <- keras_model_sequential() %>%
  layer_lstm(units = 32, input_shape = dim(X_train)[2:3]) %>%
  layer_dense(units = 1)
lstm_model %>% compile(optimizer = optimizer_adam(lr = 1e-3), loss = "mse")
history_lstm <- lstm_model %>% fit(X_train, y_train, epochs = 10, batch_size = 32,
                                   validation_data = list(X_val, y_val))
y_pred_lstm <- lstm_model %>% predict(X_val)

# 6. Plot Predictions on Validation Set
t_val <- 1:length(y_val)
plot(t_val, y_val, type = "l", col = "black", lwd = 1.5, xlab = "Sample", ylab = "Normalized Temperature",
     main = "Model Predictions on Validation Set")
lines(t_val, y_pred_baseline, col = "blue")
lines(t_val, y_pred_dense, col = "red")
lines(t_val, y_pred_rnn, col = "green")
lines(t_val, y_pred_lstm, col = "magenta")
legend("topright", legend = c("True", "Baseline", "Dense", "RNN", "LSTM"),
       col = c("black", "blue", "red", "green", "magenta"), lty = 1)
